fdbclient/ThreadSafeTransaction.cpp

```diff
@@ -284,6 +285,107 @@ ThreadFuture<Standalone<VectorRef<const char*>>> ThreadSafeTransaction::getAddre
 	});
 }
 
+ThreadFuture<Standalone<VectorRef<KeyRangeRef>>> ThreadSafeTransaction::getBlobGranuleRanges(
+    const KeyRangeRef& keyRange) {
+	ISingleThreadTransaction* tr = this->tr;
+	KeyRange r = keyRange;
+
+	return onMainThread([tr, r]() -> Future<Standalone<VectorRef<KeyRangeRef>>> {
+		tr->checkDeferredError();
+		return tr->getBlobGranuleRanges(r);
+	});
+}
+
+ThreadFuture<RangeResult> ThreadSafeTransaction::readBlobGranules(const KeyRangeRef& keyRange,
+                                                                  Version beginVersion,
+                                                                  Optional<Version> readVersion,
+                                                                  ReadBlobGranuleContext granule_context) {
+	// In V1 of api this is required, field is just for forward compatibility
+	ASSERT(beginVersion == 0);
+
+	bool doMaterialize = !granule_context.debugNoMaterialize;
+
+	ISingleThreadTransaction* tr = this->tr;
+	KeyRange r = keyRange;
+
+	int64_t readVersionOut;
+	ThreadFuture<Standalone<VectorRef<BlobGranuleChunkRef>>> getFilesFuture = onMainThread(
+	    [tr, r, beginVersion, readVersion, &readVersionOut]() -> Future<Standalone<VectorRef<BlobGranuleChunkRef>>> {
+		    tr->checkDeferredError();
+		    return tr->readBlobGranules(r, beginVersion, readVersion, &readVersionOut);
+	    });
+
+	// FIXME: can this safely avoid another main thread jump?
+	getFilesFuture.blockUntilReadyCheckOnMainThread();
```

Comment:
I think it's possible this could be called from the main thread, and we may not want to treat that as an error. Instead, you may be able to use something like `mapThreadFuture` to schedule some work to happen when the first future completes.